e2e/admin_test.go

```diff
@@ -61,23 +44,40 @@ func (c *client) isReady() (bool, error) {
 	return resp.StatusCode == http.StatusOK, nil
 }
 
-func (c *client) getStats() (*stats, error) {
-	resp, err := http.Get(c.baseURL + "/stats?format=json")
-	if err != nil {
-		return nil, err
+func (c *adminClient) getMainListenerURL() (string, error) {
+	var s map[string]interface{}
+	if err := c.getJSON("/listeners", &s); err != nil {
+		return "", err
 	}
-	defer resp.Body.Close() //nolint:errcheck
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+
+	// The json structure is deep, so parsing instead of many nested structs
+	for _, s := range s["listener_statuses"].([]interface{}) {
+		l := s.(map[string]interface{})
+		if l["name"] != "main" {
+			continue
+		}
+		port := l["local_address"].(map[string]interface{})["socket_address"].(map[string]interface{})["port_value"]
+		return fmt.Sprintf("http://127.0.0.1:%d", int(port.(float64))), nil
 	}
-	body, err := io.ReadAll(resp.Body)
+	return "", fmt.Errorf("didn't find main port in %+v", s)
+}
+
+func (c *adminClient) getJSON(path string, v interface{}) error {
+	body, err := get(c.baseURL + path + "?format=json")
 	if err != nil {
-		return nil, err
+		return err
 	}
-	var s stats
-	err = json.Unmarshal(body, &s)
+	return json.Unmarshal(body, v)
+}
+
+func get(url string) ([]byte, error) {
```

Comment:
nit want to prefix http.. 😄 
```suggestion
func httpGet(url string) ([]byte, error) {
```